external-crates/move/crates/move-package-alt/src/errors.rs

```diff
@@ -4,14 +4,229 @@
+    },
+
+    /// Invalid environment specification
+    #[error("Invalid environment '{name}': {message}")]
+    InvalidEnvironment {
+        name: String,
+        message: String,
+        span: Option<Range<usize>>,
+        path: PathBuf,
+        source_code: String,
+    },
+
+    /// Invalid dependency override
+    #[error("Invalid dependency override for '{name}' in environment '{env}': {message}")]
+    InvalidDependencyOverride {
+        name: String,
+        env: String,
+        message: String,
+        span: Option<Range<usize>>,
+        path: PathBuf,
+        source_code: String,
+    },
+
+    /// Unknown fields in the manifest
+    #[error("Unknown fields in manifest: {fields:?}")]
+    UnknownFields {
+        fields: Vec<String>,
+        span: Option<Range<usize>>,
+        path: PathBuf,
+        source_code: String,
+    },
+
+    /// Deserialization error from toml_edit
+    #[error("Failed to parse manifest: {message}")]
+    ParseError {
+        message: String,
+        span: Option<Range<usize>>,
+        path: PathBuf,
+        source_code: String,
+    },
+}
+
+impl PackageError {
+    pub fn to_diagnostic(&self) -> Diagnostic<usize> {
+        match self {
+            Self::Manifest(err) => err.to_diagnostic(),
+            _ => Diagnostic::error()
+                .with_message(self.to_string())
+                .with_labels(vec![]),
+        }
+    }
+
+    pub fn emit(&self) -> Result<(), codespan_reporting::files::Error> {
+        match self {
+            Self::Manifest(err) => err.emit(),
+            _ => Ok(()),
+        }
+    }
+}
+
+impl ManifestError {
+    /// Convert this error into a codespan Diagnostic
+    pub fn to_diagnostic(&self) -> Diagnostic<usize> {
+        let (file_id, span) = self.span_info();
+
+        match self {
+            Self::InvalidPackageName { message, .. } => Diagnostic::error()
+                .with_message(message)
+                .with_labels(vec![Label::primary(file_id, span.unwrap_or_default())]),
+            Self::InvalidEdition { message, .. } => Diagnostic::error()
+                .with_message(message)
+                .with_labels(vec![Label::primary(file_id, span.unwrap_or_default())]),
+            Self::InvalidDependency { name, message, .. } => Diagnostic::error()
+                .with_message(format!("Invalid dependency '{}': {}", name, message))
+                .with_labels(vec![Label::primary(file_id, span.unwrap_or_default())]),
+            Self::InvalidEnvironment { name, message, .. } => Diagnostic::error()
+                .with_message(format!("Invalid environment '{}': {}", name, message))
+                .with_labels(vec![Label::primary(file_id, span.unwrap_or_default())]),
+            Self::InvalidDependencyOverride {
+                name, env, message, ..
+            } => Diagnostic::error()
+                .with_message(format!(
+                    "Invalid dependency override for '{}' in environment '{}': {}",
```

Comment:
It would be nice to get all the strings together. I'm not sure how possible that is if we want multipart messages with `this_error`, but I'd like to understand how close we can get.